InformationScripting/src/graph/Graph.cpp

```diff
@@ -0,0 +1,191 @@
+** following conditions are met:
+**
+**    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+**      disclaimer.
+**    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+**      following disclaimer in the documentation and/or other materials provided with the distribution.
+**    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+**      derived from this software without specific prior written permission.
+**
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "Graph.h"
+
+#include "InformationNode.h"
+#include "InformationEdge.h"
+
+namespace InformationScripting {
+
+QList<Graph::NodeHash> Graph::nodeHashFunctions_;
+
+Graph::~Graph()
+{
+	for (auto& e : edges_) SAFE_DELETE(e);
+	for (auto& n : nodes_) SAFE_DELETE(n);
+}
+
+InformationNode* Graph::add(InformationNode* node)
+{
+	Q_ASSERT(!nodeHashFunctions_.empty());
+
+	InformationNode* existingNode = findNode(node);
+
+	if (existingNode && existingNode != node)
+	{
+		mergeNodes(existingNode, node);
+		SAFE_DELETE(node);
+	}
+	else
+	{
+		nodes_[hashValueOf(node)] = node;
+		existingNode = node;
+	}
+	return existingNode;
+}
+
+InformationEdge* Graph::addDirectedEdge(InformationNode* from, InformationNode* to, const QString& name)
+{
+	// We only allow existing nodes:
+	Q_ASSERT(from == findNode(from));
+	Q_ASSERT(to == findNode(to));
+	for (auto edge : edges_)
+	{
+		if (edge->from() == from && edge->to() == to && edge->name() == name)
+		{
+			auto existingEdge = edge;
+			Q_ASSERT(existingEdge->isDirected());
+			existingEdge->incrementCount();
+			return existingEdge;
+		}
+	}
+	auto edge = new InformationEdge(from, to, name);
+	edges_.push_back(edge);
+	return edge;
+}
+
+InformationEdge* Graph::addEdge(InformationNode* a, InformationNode* b, const QString& name)
+{
+	// We only allow existing nodes:
+	Q_ASSERT(a == findNode(a));
```

Comment:
This function is almost identical to the one above. We should have only one function `addEdge` (I know I said otherwise before, but that's before I saw the implementations). This one function should take an enum as the last argument that is either Directed or Undirected, and should have a default value (pick one, I'm not sure what will be more common).